presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapColumnWriter.java

```diff
@@ -0,0 +1,379 @@
+        this.sequence = sequence;
+        this.keyColumnId = orcTypes.get(column).getFieldTypeIndex(0);
+        this.valueColumnId = orcTypes.get(column).getFieldTypeIndex(1);
+
+        this.type = (MapType) type;
+        this.keyType = type.getTypeParameters().get(0);
+        this.valueType = type.getTypeParameters().get(1);
+
+        this.keyOrcType = orcTypes.get(keyColumnId);
+        this.valueOrcType = orcTypes.get(valueColumnId);
+        this.compressed = columnWriterOptions.getCompressionKind() != NONE;
+
+        this.columnEncoding = new ColumnEncoding(DWRF_MAP_FLAT, 0);
+        this.keyColumnEncoding = new ColumnEncoding(DIRECT, 0);
+
+        if (this.valueOrcType.getOrcTypeKind() == MAP) {
+            this.valueColumnEncoding = new ColumnEncoding(DWRF_MAP_FLAT, 0);
+        }
+        else {
+            this.valueColumnEncoding = new ColumnEncoding(DIRECT, 0);
+        }
+
+        this.dwrfEncryptor = dwrfEncryptor;
+        this.metadataWriter = new CompressedMetadataWriter(metadataWriter, columnWriterOptions, dwrfEncryptor);
+        this.inputMetadataWriter = metadataWriter;
+
+        this.valueWriterFactory = valueWriterfactory;
+        this.stringStatisticsLimitInBytes = toIntExact(columnWriterOptions.getStringStatisticsLimit().toBytes());
+
+        this.lengthStream = createLengthOutputStream(columnWriterOptions, dwrfEncryptor, orcEncoding);
+        this.presentStream = new PresentOutputStream(columnWriterOptions, dwrfEncryptor);
+        this.keyNodes = new ArrayList<>();
+
+        switch (keyOrcType.getOrcTypeKind()) {
+            case CHAR:
+            case VARCHAR:
+            case STRING:
+                this.blockWriter = new FlatMapStringKeyBlockWriter(this.valueColumnId, columnWriterOptions, dwrfEncryptor, inputMetadataWriter, type, this.keyOrcType, this.valueWriterFactory, this.keyNodes);
+                this.statisticsBuilderSupplier = this::newStringStatisticsBuilder;
+                break;
+            case BYTE:
+            case SHORT:
+            case INT:
+            case LONG:
+                this.blockWriter = new FlatMapNumericKeyBlockWriter(this.valueColumnId, columnWriterOptions, dwrfEncryptor, inputMetadataWriter, type, this.keyOrcType, this.valueWriterFactory, this.keyNodes);
+                this.statisticsBuilderSupplier = IntegerStatisticsBuilder::new;
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported type for FlatMap Keys : " + type);
+        }
+        this.statisticsBuilder = statisticsBuilderSupplier.get();
+    }
+
+    @Override
+    public List<ColumnWriter> getNestedColumnWriters()
+    {
+        return ImmutableList.of();
+    }
+
+    @Override
+    public Map<Integer, ColumnEncoding> getColumnEncodings()
+    {
+        ImmutableMap.Builder<Integer, ColumnEncoding> encodings = ImmutableMap.builder();
+
+        encodings.put(column, columnEncoding);
+        encodings.put(keyColumnId, keyColumnEncoding);
```

Comment:
do we write keyEncoding for Flat Map ? I thought we only wrote the map and the value encodings.